File: src/backend/llvm/guards.py
"""Trap guards realising the CIR semantics on LLVM.

Module M5.  Owner: Member 3.
Status: IMPLEMENTED for rows 1, 2, 3 and 6 of docs/divergence.md; row 7
        (array bounds) is here too, driven by the extents emit_ll.py recovers
        from each ``alloca``.

LLVM leaves undefined what CIR defines, so each function here emits the code
that closes one row of the divergence table.  This is the asymmetry the report
turns on: WebAssembly traps natively, so almost all the guard code lands on
the LLVM side.  Without it, the same CIR would abort in a browser and run on
to a garbage result natively -- which is exactly the bug the project exists to
prevent.

  row 1  sdiv/srem/udiv/urem by zero  -> compare + branch to a trap block
  row 2  INT_MIN / -1                 -> second condition on the same branch
  row 3  shl/ashr/lshr count >= width -> mask the count to width-1
  row 6  fptosi out of range or NaN   -> range and NaN check before the cast
  row 7  array index out of bounds    -> unsigned compare against the extent

Emitter protocol
----------------
Each function takes the emit_ll.Emitter as its first argument and uses four
methods of it -- ``tmp()``, ``label()``, ``line()`` and ``trap_branch()`` --
so that the guards stay independent of how the emitter buffers its output.
"""

from __future__ import annotations

from ...cir.ir import ConstInt, INT_WIDTH, Ty, wrap_int
from .typemap import llvm_ty

# The most negative value of each signed integer type, whose division by -1
# overflows and is therefore undefined in LLVM (docs/divergence.md row 2).
INT_MIN = {ty: -(1 << (bits - 1)) for ty, bits in INT_WIDTH.items()}


def guard_div(em, op: str, ty: Ty, lhs: str, rhs, rhs_text: str) -> None:
    """Emit the divide/remainder guard in front of an sdiv/udiv/srem/urem.

    ``rhs`` is the CIR operand, so a constant divisor can be settled at
    compile time: any constant other than 0 or -1 provably cannot trap and
    needs no run-time test at all.  0 and -1 keep the full guard rather than
    being special-cased into an unconditional trap, because that keeps one
    code path to reason about.
    """
    if isinstance(rhs, ConstInt) and wrap_int(rhs.value, ty) not in (0, -1):
        return

    lty = llvm_ty(ty)
    signed = op in ("sdiv", "srem")

    zero = em.tmp()
    em.line(f"{zero} = icmp eq {lty} {rhs_text}, 0")
    bad = zero

    if signed:
        min_lhs = em.tmp()
        minus_one = em.tmp()
        overflow = em.tmp()
        combined = em.tmp()
        em.line(f"{min_lhs} = icmp eq {lty} {lhs}, {INT_MIN[ty]}")
        em.line(f"{minus_one} = icmp eq {lty} {rhs_text}, -1")
        em.line(f"{overflow} = and i1 {min_lhs}, {minus_one}")
        em.line(f"{combined} = or i1 {zero}, {overflow}")
        bad = combined

    em.trap_branch(bad, "div")


def mask_shift_count(em, ty: Ty, count, count_text: str) -> str:
    """Return the LLVM operand to shift by, with the count reduced mod width.

    CIR defines ``x << 33`` as ``x << 1`` (docs/divergence.md row 3) whereas
    LLVM makes an over-wide shift poison.  A constant count is masked here so
    the emitted IR stays readable; a dynamic count costs one ``and``, which is
    also exactly what a C compiler emits for the same reason.
    """
    mask = INT_WIDTH[ty] - 1
    if isinstance(count, ConstInt):
        return str(wrap_int(count.value & mask, ty))
    masked = em.tmp()
    em.line(f"{masked} = and {llvm_ty(ty)} {count_text}, {mask}")
    return masked


def guard_fptosi(em, ty: Ty, value_text: str) -> None:
    """Emit the NaN and range check in front of an fptosi.

    Truncation toward zero is defined for exactly those doubles strictly
    inside (INT_MIN-1, INT_MAX+1); anything else, NaN included, is undefined
    in LLVM and a trap in CIR (docs/divergence.md row 6).  Both bounds are
    powers of two plus one, so both are exact doubles and the comparison is
    not itself a rounding question.
    """
    bits = INT_WIDTH[ty]
    low = -(1 << (bits - 1)) - 1          # one below INT_MIN
    high = 1 << (bits - 1)                # one above INT_MAX

    is_nan = em.tmp()
    too_low = em.tmp()
    too_high = em.tmp()
    out_of_range = em.tmp()
    bad = em.tmp()
    em.line(f"{is_nan} = fcmp uno double {value_text}, {value_text}")
    em.line(f"{too_low} = fcmp ole double {value_text}, {float(low)}")
    em.line(f"{too_high} = fcmp oge double {value_text}, {float(high)}")
    em.line(f"{out_of_range} = or i1 {too_low}, {too_high}")
    em.line(f"{bad} = or i1 {is_nan}, {out_of_range}")
    em.trap_branch(bad, "fptosi")


def guard_bounds(em, extent: int, index, index_text: str) -> None:
    """Emit the array bounds check in front of a gep.

    An *unsigned* comparison against the extent rejects a negative index in
    the same instruction, because a negative i32 is a very large u32.  CIR
    traps on an out-of-bounds index (docs/divergence.md row 7); LLVM would
    happily compute the address and read whatever is there.
    """
    if isinstance(index, ConstInt) and 0 <= index.value < extent:
        return                                  # provably in bounds
    bad = em.tmp()
    em.line(f"{bad} = icmp uge i32 {index_text}, {extent}")
    em.trap_branch(bad, "bounds")

File: src/backend/llvm/Guards.cpp
#include "mtir/backend/llvm/Guards.h"

#include <cstdint>
#include <string>

#include "mtir/backend/llvm/TypeMap.h"
#include "mtir/cir/Arith.h"
#include "mtir/cir/Printer.h"

namespace mtir::backend::llvm {
namespace {

using cir::ConstInt;
using cir::Opcode;
using cir::Ty;
using cir::Value;

std::string ty(Ty t) { return std::string(llvmType(t)); }

bool isSignedDivision(Opcode op) {
  return op == Opcode::SDiv || op == Opcode::SRem;
}

} // namespace

void guardDivision(Emitter &em, Opcode op, Ty t, std::string_view lhs,
                   const Value &rhs, std::string_view rhsText) {
  const bool isSigned = isSignedDivision(op);

  if (const ConstInt *c = cir::asConstInt(rhs)) {
    const std::int64_t divisor = cir::wrapInt(c->value, t);
    if (divisor != 0 && divisor != -1)
      return; // provably cannot trap
  }

  const std::string lty = ty(t);
  const std::string zero = em.tmp();
  em.line(zero + " = icmp eq " + lty + " " + std::string(rhsText) + ", 0");
  std::string bad = zero;

  if (isSigned) {
    // row 2: INT_MIN / -1 overflows, which LLVM also leaves undefined.
    const std::string minLhs = em.tmp();
    const std::string minusOne = em.tmp();
    const std::string overflow = em.tmp();
    const std::string combined = em.tmp();
    em.line(minLhs + " = icmp eq " + lty + " " + std::string(lhs) + ", " +
            std::to_string(cir::intMin(t)));
    em.line(minusOne + " = icmp eq " + lty + " " + std::string(rhsText) + ", -1");
    em.line(overflow + " = and i1 " + minLhs + ", " + minusOne);
    em.line(combined + " = or i1 " + zero + ", " + overflow);
    bad = combined;
  }

  em.trapBranch(bad, "div");
}

std::string maskShiftCount(Emitter &em, Ty t, const Value &count,
                           std::string_view countText) {
  // A shift on a non-integer type is malformed CIR, which emitModule rejects
  // before reaching here.  The check is kept anyway because the alternative is
  // `intWidth(t) - 1` underflowing to 0xFFFFFFFF and masking with garbage.
  const unsigned bits = cir::intWidth(t);
  if (bits == 0)
    return std::string(countText);

  const unsigned mask = bits - 1;
  if (const ConstInt *c = cir::asConstInt(count))
    return std::to_string(cir::wrapInt(cir::maskShift(c->value, t), t));

  const std::string masked = em.tmp();
  em.line(masked + " = and " + ty(t) + " " + std::string(countText) + ", " +
          std::to_string(mask));
  return masked;
}

void guardFPToSI(Emitter &em, Ty t, std::string_view valueText) {
  const unsigned bits = cir::intWidth(t);
  // An fptosi whose result type is not an integer is malformed CIR, which
  // emitModule rejects before reaching here.  Without this check, `bits - 1`
  // underflows and the shift below is undefined behaviour -- which is exactly
  // the class of bug docs/divergence.md exists to keep out of this compiler.
  if (bits == 0)
    return;

  // Truncation toward zero is defined exactly for
  //     -(2^(bits-1)) - 1  <  x  <  2^(bits-1)
  // Both 2^(bits-1) and -(2^(bits-1)) are exact doubles for every width CIR
  // has.  The *lower* bound is the subtle one: -(2^(bits-1)) - 1 is exact for
  // i32 (-2147483649) but NOT for i64, where it rounds back to -(2^63).  Using
  // `ole -(2^63)` there would trap on x = -2^63, which is a perfectly valid
  // conversion.  So when the exclusive bound is not representable, compare
  // strictly below the inclusive one instead.
  const double minValue = -static_cast<double>(std::uint64_t{1} << (bits - 1));
  const double high = static_cast<double>(std::uint64_t{1} << (bits - 1));
  const double lowExclusive = minValue - 1.0;
  const bool lowIsExact = lowExclusive != minValue;

  const std::string v(valueText);

  const std::string isNan = em.tmp();
  const std::string tooLow = em.tmp();
  const std::string tooHigh = em.tmp();
  const std::string outOfRange = em.tmp();
  const std::string bad = em.tmp();

  em.line(isNan + " = fcmp uno double " + v + ", " + v);
  em.line(tooLow + " = fcmp " + (lowIsExact ? "ole " : "olt ") + "double " + v +
          ", " + cir::printDouble(lowIsExact ? lowExclusive : minValue));
  em.line(tooHigh + " = fcmp oge double " + v + ", " + cir::printDouble(high));
  em.line(outOfRange + " = or i1 " + tooLow + ", " + tooHigh);
  em.line(bad + " = or i1 " + isNan + ", " + outOfRange);
  em.trapBranch(bad, "fptosi");
}

void guardBounds(Emitter &em, std::int64_t extent, const Value &index,
                 std::string_view indexText) {
  if (const ConstInt *c = cir::asConstInt(index))
    if (c->value >= 0 && c->value < extent)
      return; // provably in bounds

  const std::string bad = em.tmp();
  em.line(bad + " = icmp uge i32 " + std::string(indexText) + ", " +
          std::to_string(extent));
  em.trapBranch(bad, "bounds");
}

} // namespace mtir::backend::llvm
